_posts/2024-12-xx-hybrid-search-optimization.md

```diff
@@ -0,0 +1,342 @@
+
+### Query features
+
+* Number of terms: how many terms does the user query have?  
+* Query length: how long is the user query (measured in characters)?  
+* Contains number: does the query contain one or more numbers?  
+* Contains special character: does the query contain one or more special characters (non-alphanumeric characters)?
+
+### Lexical search result features
+
+* Number of results: the number of results for the lexical query.  
+* Maximum title score: the maximum score of the titles of the retrieved top 10 documents. The scores are BM25 scores calculated individually per result set. That means that the BM25 score is not calculated on the whole index but only on the retrieved subset for the query, making the scores more comparable to each other and less prone to outliers that could result from high IDF values for very rare query terms.  
+* Sum of title scores: the sum of the title scores of the top 10 documents, again calculated per-result set. We use the sum of the scores (and no average value) as an aggregate to have a measure of how relevant all retrieved top 10 titles are. BM25 scores are not normalized so using the sum instead of the average seemed reasonable.
+
+### Neural search result features
+
+* Maximum semantic score: the maximum semantic score of the retrieved top 10 documents. This is the score we receive for a neural query based on the query’s similarity to the title.  
+* Average semantic score: By contrast to BM 25 scores the semantic scores are normalized and in the range of 0 to 1\. Using the average score seems more reasonable than going for the sum here.
+
+## Feature engineering
+
+As training data we used the output of the global hybrid search optimizer. As part of this process we ran every query 66 times: once per hybrid search configuration. For each query we calculated the search metrics and as a result we know per query which pipeline worked best, thus also which “neuralness” (neural search weight) worked best. We used the best NDCG@10 value per query as the metric deciding what the ideal “neuralness” was.
+
+That leaves us with 250 queries (small query set) or 5,000 queries (large query set) together with their “neuralness” values for which they achieved best NDCG@10 values. Next, we engineered the nine features for each query. This constitutes the training and test data.
+
+## Model training and evaluation
+
+With the appropriate data at hand we explored different algorithms and experimented with different model fitting settings to identify patterns and evaluate if we’re on the right track with that approach.  
+We went for two relatively simple algorithms: linear regression and random forest regression.  
+We applied cross validation, regularization, and tried out all different feature combinations. This resulted in interesting findings that are summarized in the following section.
+
```

Comment:
Because we say "in the following section", should there be an H3 here like " ### Findings"?

Reply:
I don't see an extra H3 heading , so I'm changing the heading of this paragraph to "## Model training and evaluation findings" (added "findings")